Add Gulp to project which improves tooling, remove .iml file

ebramanti opened the pull request:
This PR is to address #1, which actually requires the user to have Node, install Jade globally and have Python installed. I've removed the global Jade and Python dependency requirements, and instead opted for local dependency installation. 

I've also added Gulp to the project, which uses a few plugins to statically serve files and watch Jade/JS/Datasource changes just like the old scripts. One of the packages I'd like to point out in particular is `gulp-connect`, which adds livereload functionality! When you save any of your project files, the page automatically reloads and the developer can see their changes instantly.

I've also snuck in an unrelated chore: deleted an iml file that is committed, but the `.gitignore` missed.

Let me know what you think @AuthorOfTheSurf!

ebramanti committed:
Add Gulp to project with improved tooling

- Remove necessity for global install of Jade
- Remove Python dependency
- Use Gulp since Node is already a dependency
- Add live reload to project

ebramanti committed:
Remove iml file from project

Is gitignored, but not yet removed

jadengore commented on compile-jade.sh:
```diff
@@ -1,24 +0,0 @@
-# rather than watching an entire directory. Edits to the data source,
-# i.e. json file backing a jade file will not trigger the recompile.
-# In that case you will want to cancel, and re-run this script to force
-# a recompile.
```
Also, to address your comment here in the future we can incorporate this package: https://www.npmjs.com/package/gulp-jade-find-affected

AuthorOfTheSurf commented on Gulpfile.js:
```diff
@@ -0,0 +1,48 @@
+gulp.task('jade', () => {
+  gulp.src('jade/*.jade')
+    .pipe(jade({
+      locals: JSON.parse(fs.readFileSync('./data-source.json', 'utf8')),
```
1) The current Jade watcher allows the user to specify a particular json data source for each Jade template. It looks like this will use the same data source for every template, or will require the user to copy this 8-line gulp pipeline for each template-datasource pair. Can you make this a parameterized function? e.g. `watchAndRenderWithData(jadeTemplate, jsonDataFile)`

2) I intend to put data sources in a top-level `data/` directory, so that change is incoming.

ebramanti replied:
1) We can specify this as a command line option when you run `npm run gulp`!

2) If you put everything in a data directory, we can default all of those to be part of `locals`, with any optional parameters included.

AuthorOfTheSurf commented on Gulpfile.js:
```diff
@@ -0,0 +1,48 @@
+
+  gulp.src('index.jade')
+    .pipe(jade())
+    .pipe(gulp.dest('build'));
```
I like the use of a `build/` directory

AuthorOfTheSurf commented on Gulpfile.js:
```diff
@@ -0,0 +1,48 @@
+});
+
+gulp.task('watch-data-source', () => {
+  gulp.watch('data-source.json', ['jade']);
```
Nice. One problem with `jade --watch` is that it only responds to a Jade template (or related template) change, not to changes to json data sources. This looks like a nice solution to that!

As per above, it would be really nice to parameterize this. So when the user adds a `watchAndRenderWith(...)` entry, this `gulp.watch` action will be done on the data source.

AuthorOfTheSurf commented on README.md:
```diff
@@ -2,12 +2,14 @@
-6. Visit pages at `localhost:8000`
-7. Take a look at source code to see how to get Jade and Knockout.js working together given a single data-blob as a starting point.
+3. Install dependencies: `npm install`
+4. Start static file server & Jade/JS/Datasource watch: `npm run gulp`
```
Woot!

AuthorOfTheSurf commented on package.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "jade-knockout-template",
```
Good idea. We are in node-land and I was trying to avoid a package.json by just doing stuff globally. This is the way to go as keeping deps local is nice. Being able to run the server without dipping into python-land is extra nice.

AuthorOfTheSurf commented on python-server.sh:
```diff
@@ -1,7 +0,0 @@
-# have to double-click the html files. This is closer to what the development
-# environment would be like. Requires Python 2.7
-# I recommend running this in its own terminal window/tab
-python -m SimpleHTTPServer 8000
```
Goodbye

AuthorOfTheSurf commented:
Main thing is making the gulp actions more parameterized, e.g. imagine 10 different `.jade` files with 10 different `.json` data source files. This should be doable by having 10 `watchAndRenderWith(...)` function calls, i.e. not copy-pasting gulp stuff.

I also want to put json data in a `data/` directory... could you include this in your PR :)?

Thanks for the PR and JS wisdom!

ebramanti committed:
Update jade pipeline to parse from data folder

Jade files will now have a matching file in the /data folder (will be
created if it DNE), and they will be interpolated in general data
rather than the locals object.

ebramanti committed:
Update .gitignore

ebramanti commented:
@AuthorOfTheSurf Updated with a `data` directory, the compiler will now create a corresponding JSON file for each Jade template automatically!

ebramanti committed:
Data paths now match directory structure

- Match directory structure in data folder
- Improved business logic for parseJSONData
- Add mkdirp to create new folders if they don’t exist

ebramanti committed:
Update README

Updates readme to include info about data directory

ebramanti commented:
@AuthorOfTheSurf ping on this

ebramanti commented:
@AuthorOfTheSurf  👀❗

AuthorOfTheSurf commented:
Thanks for the changes—lets get this merged, several helpful improvements here!

ebramanti commented:
💯
